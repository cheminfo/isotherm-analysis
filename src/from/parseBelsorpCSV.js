import { ensureString } from 'ensure-string';

const variableLabels = {
  No: 'Peak number',
  pi: 'Initial pressure', // kPa
  pe: 'Absolute pressure', // kPa
  p0: 'Relative pressure', // kPa
  'p/p0': 'p/p0',
  na: 'Amount of absorbed gas', // mol g^-1
  Va: 'Amount of absorbed gas', // ml g^-1
};

// p0 = saturation pressure
// p/p0 = relative pressure
// na = amount of adsorbed gas (N2 or CO2 for example)

/*
For adsorption we want:
- x: p/p0
- y: na or va
- p:

Adsorptive,N2
Adsorption temperature,77.000,[K]

For desorption we want
- x:
- y:
- p:


*/

export function parseBelsorpCSV(binary) {
  const text = ensureString(binary);
  const result = {};
  result.meta = {};

  const [header, data] = text.split(/\r?\n,*\r?\n/);
  appendMeta(header, result);
  appendData(data, result);

  return result;
}

function appendMeta(header, result) {
  for (let line of header.split(/\r?\n/)) {
    let comma = line.indexOf(',');
    if (comma > 0) {
      result.meta[line.substring(0, comma)] = line
        .substring(comma + 1)
        .replace(',[', ' [')
        .replace(/,+$/, '');
    }
  }
}

function appendData(data, result) {
  const lines = data.split(/\r?\n/);
  const variables = {};
  const variableNames = [];
  for (let field of lines[0].split(',')) {
    const parts = field === 'p/p0' ? [field] : field.split('/');
    const variable = {
      label: variableLabels[parts[0]] || parts[0],
      variable: parts[0],
      units: parts[1] ? parts[1].replace(' g^-1', '/g') : '',
      data: [],
    };
    variableNames.push(parts[0]);
    variables[parts[0]] = variable;
  }

  let target;
  for (let line of lines) {
    if (line.startsWith('ADS')) {
      target = JSON.parse(JSON.stringify(variables));
      result.adsorption = target;
    } else if (line.startsWith('DES')) {
      target = JSON.parse(JSON.stringify(variables));
      result.desorption = target;
    }
    if (target) {
      if (line.match(/^[0-9E,.-]*$/)) {
        const fields = line.split(',').map((field) => Number(field));
        for (let i = 0; i < variableNames.length; i++) {
          if (variableNames[i]) {
            target[variableNames[i]].data.push(fields[i]);
          }
        }
      }
    }
  }
}
