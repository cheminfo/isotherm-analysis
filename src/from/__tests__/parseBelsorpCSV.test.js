import { readFileSync } from 'fs';
import { join } from 'path';

import { parseBelsorpCSV } from '../parseBelsorpCSV';

test('parseBelsorpCSV', () => {
  const binary = readFileSync(join(__dirname, 'data/belsorp.csv'));

  const parsed = parseBelsorpCSV(binary);

  expect(Object.keys(parsed)).toStrictEqual([
    'meta',
    'adsorption',
    'desorption',
  ]);
  expect(parsed).toMatchSnapshot();
});
