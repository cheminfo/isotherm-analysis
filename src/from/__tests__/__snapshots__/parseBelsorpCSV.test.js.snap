// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseBelsorpCSV 1`] = `
Object {
  "adsorption": Object {
    "No": Object {
      "data": Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
      ],
      "label": "Peak number",
      "units": "",
      "variable": "No",
    },
    "na": Object {
      "data": Array [
        0.000022289,
        0.000044887,
        0.00010545,
        0.00025223,
        0.00038821,
        0.00040127,
        0.00041572,
        0.00043709,
        0.00044646,
        0.00045753,
        0.00046899,
        0.00047417,
        0.00048363,
        0.00048831,
        0.00049304,
        0.00049674,
        0.00050272,
        0.00051142,
        0.00052507,
        0.00055171,
        0.00060275,
        0.00066873,
        0.00059409,
        0.00062153,
        0.00064272,
        0.0006607,
        0.00068386,
        0.00071687,
        0.00076635,
        0.00083787,
        0.0009399,
        0.0010829,
        0.0013107,
        0.0013923,
        0.0017455,
        0.0019372,
        0.0020646,
        0.0026912,
        0.0031789,
        0.003597,
        0.0042866,
        0.0048782,
        0.0054185,
        0.005928,
        0.006388,
        0.0072107,
        0.0079749,
        0.0087115,
      ],
      "label": "Amount of absorbed gas",
      "units": "mol/g",
      "variable": "na",
    },
    "p/p0": Object {
      "data": Array [
        0.00002519,
        0.000053514,
        0.00012073,
        0.00029799,
        0.0012932,
        0.0017461,
        0.0026076,
        0.0054841,
        0.0078345,
        0.010417,
        0.015887,
        0.021483,
        0.04261,
        0.058021,
        0.073734,
        0.089467,
        0.1105,
        0.1382,
        0.1682,
        0.2097,
        0.2616,
        0.3141,
        0.355,
        0.4004,
        0.4518,
        0.502,
        0.552,
        0.6016,
        0.6512,
        0.7004,
        0.7488,
        0.7965,
        0.8417,
        0.8512,
        0.8841,
        0.8946,
        0.8989,
        0.9184,
        0.927,
        0.9322,
        0.9382,
        0.9413,
        0.9436,
        0.945,
        0.9464,
        0.9485,
        0.9495,
        0.9502,
      ],
      "label": "p/p0",
      "units": "",
      "variable": "p/p0",
    },
    "p0": Object {
      "data": Array [
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
      ],
      "label": "Relative pressure",
      "units": "kPa",
      "variable": "p0",
    },
    "pe": Object {
      "data": Array [
        0.002519,
        0.0053514,
        0.012073,
        0.029799,
        0.1293,
        0.1746,
        0.2608,
        0.5484,
        0.7834,
        1.0417,
        1.5887,
        2.1483,
        4.261,
        5.8021,
        7.3734,
        8.9467,
        11.048,
        13.816,
        16.818,
        20.972,
        26.156,
        31.412,
        35.5,
        40.04,
        45.182,
        50.202,
        55.204,
        60.158,
        65.121,
        70.04,
        74.882,
        79.653,
        84.165,
        85.118,
        88.414,
        89.455,
        89.89,
        91.837,
        92.696,
        93.221,
        93.815,
        94.131,
        94.36,
        94.503,
        94.641,
        94.853,
        94.954,
        95.017,
      ],
      "label": "Absolute pressure",
      "units": "kPa",
      "variable": "pe",
    },
    "pe2": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "label": "pe2",
      "units": "kPa",
      "variable": "pe2",
    },
    "pi": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "label": "Initial pressure",
      "units": "kPa",
      "variable": "pi",
    },
  },
  "desorption": Object {
    "No": Object {
      "data": Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        0,
      ],
      "label": "Peak number",
      "units": "",
      "variable": "No",
    },
    "na": Object {
      "data": Array [
        0.0082783,
        0.0072084,
        0.0062144,
        0.0052137,
        0.0043358,
        0.0036074,
        0.0029867,
        0.002568,
        0.0018215,
        0.0014289,
        0.00134,
        0.0010485,
        0.0010128,
        0.00086486,
        0.0007548,
        0.00068038,
        0.00063005,
        0.00060302,
        0.00058049,
        0.00056955,
        0.00054921,
        0.00051936,
        0.00048032,
        0.00043257,
        0.0004195,
        0.00041188,
        0.00041282,
        0.00041751,
        0.00042107,
        0.000423,
        0.00042577,
        0.00042812,
        0.00042641,
        0.00042234,
        undefined,
      ],
      "label": "Amount of absorbed gas",
      "units": "mol/g",
      "variable": "na",
    },
    "p/p0": Object {
      "data": Array [
        0.9453,
        0.9401,
        0.9355,
        0.9302,
        0.9242,
        0.9171,
        0.91,
        0.9023,
        0.881,
        0.8572,
        0.8491,
        0.8041,
        0.7963,
        0.7502,
        0.6992,
        0.6484,
        0.5981,
        0.5478,
        0.4981,
        0.4483,
        0.3985,
        0.3488,
        0.2991,
        0.2368,
        0.2064,
        0.1515,
        0.1229,
        0.099338,
        0.080461,
        0.066185,
        0.051949,
        0.037806,
        0.02102,
        0.014369,
        undefined,
      ],
      "label": "p/p0",
      "units": "",
      "variable": "p/p0",
    },
    "p0": Object {
      "data": Array [
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        100,
        undefined,
      ],
      "label": "Relative pressure",
      "units": "kPa",
      "variable": "p0",
    },
    "pe": Object {
      "data": Array [
        94.527,
        94.01,
        93.549,
        93.019,
        92.425,
        91.709,
        91.004,
        90.233,
        88.101,
        85.715,
        84.913,
        80.405,
        79.629,
        75.024,
        69.916,
        64.839,
        59.81,
        54.78,
        49.808,
        44.826,
        39.845,
        34.884,
        29.909,
        23.679,
        20.636,
        15.151,
        12.288,
        9.9338,
        8.0461,
        6.6185,
        5.1949,
        3.7806,
        2.102,
        1.4369,
        undefined,
      ],
      "label": "Absolute pressure",
      "units": "kPa",
      "variable": "pe",
    },
    "pe2": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        undefined,
      ],
      "label": "pe2",
      "units": "kPa",
      "variable": "pe2",
    },
    "pi": Object {
      "data": Array [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        undefined,
      ],
      "label": "Initial pressure",
      "units": "kPa",
      "variable": "pi",
    },
  },
  "meta": Object {
    "Adsorption temperature": "77.000 [K]",
    "Adsorptive": "N2",
    "Adsorptive cross section area": "0.1620 [nm^2]",
    "Apparatus temperature": "0.0000 [C]",
    "COMMENT1": "Standard 1",
    "COMMENT2": "dsun",
    "COMMENT3": "",
    "COMMENT4": ", Vacuum degree before measurement:1.111E-4Pa",
    "Date of measurement": "18/02/06",
    "Dead volume": "30.402 [cm^3]",
    "Equilibrium time": "0 [SEC]",
    "File Name": "Act Carbon 2018 A.DAT",
    "File name of wall adsorption": "",
    "Number of adsorption data point :": "48",
    "Number of desorption data point :": "34",
    "Sample weight": "8.1300E-02 [g]",
    "Saturated vapor pressure": "100.000 [kPa]",
    "Serial number": "00463",
    "Standard volume": "24.502 [cm^3]",
    "Time of measurement": "30:25:37",
    "Version": "",
    "Wall adsorption correction value1": "",
    "Wall adsorption correction value2": "",
  },
}
`;
